(DOCSP-8761): Port auth provider: Email/Password (#29)
mongodb/docs-realm: source/authentication/email-password.txt

.. _email-password-authentication:

=============================
Email/Password Authentication
=============================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. _android-deep-links: https://developer.android.com/training/app-links/deep-linking.html
.. _ios-universal-links: https://developer.apple.com/library/content/documentation/General/Conceptual/AppSearch/UniversalLinks.html

.. _email-password-authentication-overview:

Overview
--------

The Email/Password authentication provider allows users to register
accounts and log in using their email address and a password. Realm

> must confirm Email/Password users before it allows them to login to

Reviewer: ```suggestion
must confirm Email/Password users before it allows them to log in to
```

~~~~~~~~~~~~~~~~~~~~~~~~~

If you configure the provider to :guilabel:`Send a Confirmation Email`,
Realm sends a confirmation email to the email address provided by new
Email/Password users when they register an account. The email contains a
link to a confirmation URL that the user must visit this URL within 30
minutes to confirm that they control the email address. For an example
of how to confirm a new Email/Password identity using the Realm SDK,
see
:ref:`Confirm a New User's Email Address
<email-password-authentication-usage-confirm-a-users-email-address>`.

Configure the following settings to have Realm automatically send a
confirmation email:

- :guilabel:`Email Confirmation URL`: the base of the URL included in
  every confirmation email. Realm appends a unique ``token`` and
  ``tokenId`` to this URL as query parameters to create a unique link
  for every confirmation. To confirm the user, extract these query
  parameters from the user's unique URL and pass the token and tokenId
  to the Client SDK's ``confirmUser`` function.

- :guilabel:`Email Confirmation Subject`: the subject of the email sent
  to the user containing the confirmation URL. This value is optional:

>   if omitted, Stich uses a default subject line instead. Custom email

Reviewer: ```suggestion
  if omitted, Realm uses a default subject line instead. Custom email
```

Send a Password Reset Email
~~~~~~~~~~~~~~~~~~~~~~~~~~~

If you configure the provider to
:guilabel:`Send a Password Reset Email`, Realm sends a Password Reset
URL to a user's email address when a user requests a password reset. The
user must visit this URL within 30 minutes to confirm their intent to
change their password. For an example of how to reset the password of an
Email/Password identity using the Realm SDK, see
:ref:`Reset a User's Password
<email-password-authentication-usage-reset-a-users-password>`.

Configure the following settings to have Realm automatically send a
password reset email:

- :guilabel:`Password Reset URL`: the base of the URL included in every
  password reset email. Realm appends a unique ``token`` and
  ``tokenId`` to this URL as query parameters to create a unique link
  for every password reset. To reset the user's password, extract these
  query parameters from the user's unique URL and pass the token and
  tokenId to the Client SDK's ``resetPassword`` function.

- :guilabel:`Reset Password Email Subject`: the subject line of the
  email that Realm sends to users when they request to reset their

>   password. This value is optional: if omitted, Stich uses a default

Reviewer: ```suggestion
  password. This value is optional: if omitted, Realm uses a default
```

Author: Misspelled Stitch thrwarted my find&replace. Shucks.

- Send custom password reset emails from a specific domain with a
  particular template using an external service.

- Interface with a MongoDB Atlas collection to implement a password
  reset "cooldown period" to prevent too many password reset
  attempts on a single account in a particular time range.

- Send custom password reset messages through a service other than
  email.

Configure the following settings to have Realm automatically run a
password reset function:

- :guilabel:`Function`: the function run when an instance of the client
  SDK calls the ``callResetPasswordFunction`` function. This function
  must return an object containing a ``status`` key mapping to a string
  with one of the following values: ``success``, ``pending``, or
  ``fail``.

.. note::

   If you configure the provider to use a password reset function then
   the client SDK ``sendResetPasswordEmail()`` function will return an
   error. You may still need to call ``resetPassword()`` if the password
   reset function returns a ``pending`` status.

The custom password reset function signature is
:wikipedia:`variadic <Variadic_function_>`: it accepts any number of

> parameters. The first of these parameters is always an object containing

Reviewer: This sentence ("The custom password ...") is overwritten... I would just remove this.

Author: I'm not sure there's any other way for a user to understand that they can pass extra parameters to this function, is there?
